### PR TITLE
Minor perthread cudalibhandles fix

### PR DESCRIPTION
**Pull request by nakul02**
Ping @niketanpansare - this patch has some very important fixes.
Could you please review?

### CODE REVIEW

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1887/

File: src/main/cpp/kernels/SystemML.cu [hunk 1]
```diff
@@ -35,12 +35,13 @@ nvcc -ptx -arch=sm_30 SystemML.cu
  */
 extern "C"
 __global__ void copy_u2l_dense(double* ret, int dim, int N) {
-	int ix = blockIdx.x * blockDim.x + threadIdx.x;
-	int iy = blockIdx.y * blockDim.y + threadIdx.y;
+	int tid = blockIdx.x * blockDim.x + threadIdx.x;
```
Comment (niketanpansare) [inline src/main/cpp/kernels/SystemML.cu]: Based on http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#compute-capabilities and java's maximum size, this makes sense for dense kernels on device with [compute capability 3.0](https://github.com/nakul02/systemml/blob/b1ef11f86650ee79cb7c2a4b61201f4d25ff048e/src/main/cpp/kernels/SystemML.cu#L23).

File: src/main/java/org/apache/sysml/runtime/instructions/gpu/context/ExecutionConfig.java [hunk 2]
```diff
@@ -89,16 +89,34 @@ public static ExecutionConfig getConfigForSimpleVectorOperations(int numCells) t
...
 	 */
-	public static ExecutionConfig getConfigForSimpleMatrixOperations(int rlen, int clen) throws DMLRuntimeException {
+	public static ExecutionConfig getConfigForMatrixOperations(int rlen, int clen) throws DMLRuntimeException {
 		int deviceNumber = 0;
 		int maxBlockDim = getMaxBlockDim(deviceNumber);
 		int blockDimX = (int) Math.min(maxBlockDim, rlen);
 		int gridDimX = (int) Math.ceil((double) rlen / blockDimX);
 		int blockDimY = (int) Math.min(Math.floor(((double) maxBlockDim) / blockDimX), clen);
 		int gridDimY = (int) Math.ceil((double) clen / blockDimY);
+		if (gridDimY > 65535)
```
Comment (niketanpansare) [inline src/main/java/org/apache/sysml/runtime/instructions/gpu/context/ExecutionConfig.java]: :+1:

File: src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUContext.java [hunk 3]
```diff
@@ -141,27 +141,46 @@ protected GPUContext(int deviceNum) throws DMLRuntimeException {
 		cudaMemGetInfo(free, total);
 
 		long start = System.nanoTime();
```
Comment (niketanpansare) [inline src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUContext.java]: can you guard this with DMLScript.Statistics

**Comment by niketanpansare** [comment]
LGTM. Thanks @nakul02  👍

**Comment by nakul02** [comment]
Thanks @niketanpansare, I shall merge.

**Pull request closed without merging**